src/psyclone/nemo.py

```diff
@@ -705,13 +607,47 @@ def match(node):
             return False
         # We are expecting something like:
         #    array(:,:,jk) = some_expression
-        # so we check the left-hand side...
+        # but we have to beware of cases like the following:
+        #   array(1,:,:) = a_func(array2(:,:,:), mask(:,:))
+        # where a_func is an array-valued function and `array2(:,:,:)`
+        # could just be `array2`.
+        # We check the left-hand side...
         lhs = node.items[0]
-        if isinstance(lhs, Fortran2003.Part_Ref):
-            colons = walk_ast(lhs.items, [Fortran2003.Subscript_Triplet])
-            if colons:
-                return True
-        return False
+        if not isinstance(lhs, Fortran2003.Part_Ref):
+            # LHS is not an array reference
+            return False
+        colons = walk_ast(lhs.items, [Fortran2003.Subscript_Triplet])
+        if not colons:
+            # LHS does not use array syntax
+            return False
+        # Now check the right-hand side...
+        rhs = node.items[2]
+        colons = walk_ast(rhs.items, [Fortran2003.Subscript_Triplet])
+        if not colons:
+            # We just have array syntax on the LHS
```

Comment:
Should that say RHS rather than LHS?

Reply:
No but it took me a while to work it out so it can't be very clear :-) I've re-worded it.